// Copyright 2017 The Ray Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";
option cc_enable_arenas = true;

package ray.rpc.autoscaler;

enum RayNodeKind {
  UNKNOWN = 0;
  HEAD = 1;
  WORKER = 2;
}

// A node type that's available for the cluster.
message InstanceType {
  // the name of the instance type. e.g. "m4.large"
  string type_name = 1;
  RayNodeKind node_kind = 2;
  // avaialble resources on the node. such as {"CPU": 4, "GPU": 1}
  map<string, double> resources = 3;
}

message GetAvailableInstanceTypesRequest {}

message GetAvailableInstanceTypesResponse {
  repeated InstanceType instance_types = 1;
  // number of instances that are available for starting.
  // this can change if the cloud provider has a limit on
  // number of instances that can be started.
  int64 available_instances = 2;
}

// Represents the state of a launched instance.
// An instance is considered launched as long as
// it has a unique instance_id associated with it.
//
// Note a launched instance may be DEAD. In this case,
// the state will be garbage collected after some timeout
// period (by default 30 minutes).
message Instance {
  enum InstanceStatus {
    UNKNOWN = 0;
    // The unspecified state - most likey it is queued.
    QEUEUD = 1;
    REQUESTED = 2;
    ALLOCATED = 3;
    ALLOCATION_FAILED = 4;
    STOPPING = 5;
    STOPPED = 6;
    GARBAGE_COLLECTED = 7;
  }

  enum RayStatus {
    RAY_STATUS_UNKOWN = 0;
    RAY_INSTALLING = 1;
    RAY_INSTALL_FAILED = 2;
    RAY_RUNNING = 3;
    RAY_FAILED = 4;
  }

  // an unique id for the instance that's generated by the
  // instance manager. This may be optional if
  // the instance hasn't be started yet.
  string instance_id = 11;
  // the status of the instance.
  InstanceStatus status = 12;
  // the node type of the instance.
  string instance_type = 13;
  // The corresponding total resources on the node.
  map<string, double> total_resources = 14;
  // timestamp of the last time the instance get modified.
  int64 timestamp_since_last_modified = 15;
  // the external id of the instance that's generated by
  // the cloud provider like AWS, GCP, etc.
  // Note this id can be reused by different instances.
  string cloud_instance_id = 16;
  // internal ip address of the instance.
  string internal_ip = 17;
  // external ip address of the instance.
  string external_ip = 18;
  // the monotonically increasing version number of the instance.
  int64 version = 19;
  // the status of the ray process on the instance.
  RayStatus ray_status = 20;
}

message UpdateInstanceManagerStateRequest {
  int64 expected_version = 1;
  repeated InstanceType new_nodes_to_start = 2;
  repeated string instance_ids_to_terminate = 3;
}

message UpdateInstanceManagerStateReply {
  bool success = 1;
  string error_message = 2;
  int64 version = 3;
}

message InstanceManagerState {
  // a monotonically increasing version number.
  // the version number is incremented whenever
  // the state is updated (either by successful adjusting request,
  // or instance state change).
  int64 version = 1;
  repeated Instance instances = 2;
}

message GetInstanceManagerStateRequest {}

message GetInstanceManagerStateReply {
  InstanceManagerState state = 1;
}

service InstanceManagerService {
  rpc GetInstanceManagerState(GetInstanceManagerStateRequest)
      returns (GetInstanceManagerStateReply);
  rpc UpdateInstanceManagerState(UpdateInstanceManagerStateRequest)
      returns (UpdateInstanceManagerStateReply);
  rpc GetAvailableInstanceTypes(GetAvailableInstanceTypesRequest)
      returns (GetAvailableInstanceTypesResponse);
}
